community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py

```diff
@@ -1732,39 +1713,34 @@ def machine_types(self, project=None):
             op = act.aggregatedList_next(op, result)
         return machines
 
-    def machine_type(self, machine_type, project=None, zone=None):
+    def machine_type(self, machine_type: str):
         """ """
         custom_patt = re.compile(
             r"((?P<family>\w+)-)?custom-(?P<cpus>\d+)-(?P<mem>\d+)"
         )
         custom_match = custom_patt.match(machine_type)
-        if zone:
-            project = project or self.project
-            machine_info = ensure_execute(
-                self.compute.machineTypes().get(
-                    project=project, zone=zone, machineType=machine_type
-                )
-            )
-        elif custom_match is not None:
+        if custom_match is not None:
             groups = custom_match.groupdict()
             cpus, mem = (groups[k] for k in ["cpus", "mem"])
             machine_info = {
                 "guestCpus": int(cpus),
                 "memoryMb": int(mem),
             }
         else:
-            machines = self.machine_types(project=project)
-            machine_info = next(iter(machines[machine_type].values()), None)
-            if machine_info is None:
+            machines = self.machine_types()
+            if machine_type not in machines:
                 raise Exception(f"machine type {machine_type} not found")
+            per_zone = machines[machine_type]
+            assert per_zone
```

Comment:
Wouldn't the if-condition above always catch per_zone remaining False?

Reply:
`machines` is `Dict[str, Dict[str, something]]`, say 
```py
machines = {
  "sewing": {
     "albania": 33,
     "bobruysk": 77
   },
   "soda": {},
}

assert "sewing" in machines # OK
assert "soda" in machines   # OK
assert machines["sewing"]   # OK - not empty dict
assert machines["soda"]     # FAIL - empty dict
```

The reason I'm using crude `assert` without explanatory message for user, is that because I expect it to never fire. By the way of construction of `machines` by `machine_types` functions, we know that  if `machine_type` is present in result -> `machine_type` is present in at least one `zone`. This is assertion of code invariant, while also useful for calming type-checkers.

Reply:
About "calming type-checkers". Consider:

```py
# zebu.py
from typing import Optional

def az(coin: int) -> Optional[str]:
  return "az" if coin % 2 else None

def gel():
  za = az(1)
  print(za.lower())
```

```sh
$ mypy zebu.py  --check-untyped-defs
zebu.py:8: error: Item "None" of "str | None" has no attribute "lower"  [union-attr]
Found 1 error in 1 file (checked 1 source file)
```

We can hint `mypy`:

```py
...
def gel():
  za = az(1)
  assert za is not None # <--
  print(za.lower())
```

```sh
$ mypy zebu.py  --check-untyped-defs
Success: no issues found in 1 source file
```

Reply:
appreciate the example here, makes sense!